packages/components/src/components/pin-code/pin-code.tsx

```diff
@@ -0,0 +1,218 @@
+    }
+    if (this.isWithinForm) {
+      this.hiddenInput = initHiddenInput(this.host, this.value, this.disabled, this.required);
+    }
+  }
+
+  public componentWillUpdate(): void {
+    if (this.isWithinForm) {
+      syncHiddenInput(this.hiddenInput, this.value, this.disabled, this.required);
+    }
+  }
+
+  public componentShouldUpdate(newVal: unknown, oldVal: unknown): boolean {
+    return hasPropValueChanged(newVal, oldVal);
+  }
+
+  public render(): JSX.Element {
+    validateProps(this, propTypes);
+    attachComponentCss(this.host, getComponentCss, this.type, this.hideLabel, this.state, this.disabled, this.theme);
+
+    // reset array of input elements
+    this.pinCodeElements = [];
+
+    return (
+      <Host>
+        <label class="label">
+          {hasLabel(this.host, this.label) && (
+            <span class="label__text">
+              {this.label || <slot name="label" />}
+              {this.required && <Required />}
+            </span>
+          )}
+          {hasDescription(this.host, this.description) && (
+            <span class="label__text">{this.description || <slot name="description" />}</span>
+          )}
+          <div class="pin-code-container" onKeyDown={this.onKeyDown} onPaste={this.onPaste} onClick={this.onClick}>
+            {this.isWithinForm && <slot name="hidden-input" />}
+            {...Array.from({ length: this.length }).map((_value, index) => (
+              <input
+                type={this.type === 'number' ? 'text' : this.type}
+                // aria-label={}
+                // aria-labelledby={}
+                // aria-describedby={}
+                autoComplete="one-time-code"
+                pattern="\d*"
+                inputMode="numeric" // get numeric keyboard on mobile
+                maxLength={1}
+                value={this.value?.[index]}
+                disabled={this.disabled}
+                required={this.required}
+                ref={(el) => this.pinCodeElements.push(el)}
+              />
+            ))}
+          </div>
+        </label>
+        {hasMessage(this.host, this.message, this.state) && (
+          <StateMessage state={this.state} message={this.message} theme="light" host={this.host} />
+        )}
+      </Host>
+    );
+  }
+
+  private onClick = (
+    e: MouseEvent & { target: HTMLInputElement & { previousElementSibling: HTMLInputElement } }
+  ): void => {
+    // only allow focus on filled inputs or the first empty input
+    if (!e.target.value && e.target.previousElementSibling && !e.target.previousElementSibling.value) {
+      this.pinCodeElements.find((pinCodeElement) => !pinCodeElement.value).focus();
+    }
+  };
+
+  private onKeyDown = (
+    e: KeyboardEvent & {
+      target: HTMLInputElement & { previousElementSibling: HTMLInputElement; nextElementSibling: HTMLInputElement };
+    }
+  ): void => {
+    // if input is valid overwrite old value
+    if (inputIsSingleDigit(e.key)) {
+      e.target.value = e.key;
```

Comment:
Probably would be good to destructure target and its properties that are read over and over again.